File: 802.find-eventual-safe-states.py
#
# @lc app=leetcode id=802 lang=python3
#
# [802] Find Eventual Safe States
#

# @lc code=start
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        width = len(graph)
        record = [0]*width  # 0 not seen, 1 pass, 2 terminal
        res = []

        def helpler(pos):
            if record[pos] != 0:
                return record[pos]
            else:
                record[pos] = 1
            for n in graph[pos]:
                if record[n] == 1:
                    return record[pos]
                elif record[n] == 0:
                    record[n] = helpler(n)
                    if record[n] == 1:
                        return record[pos]
            record[pos] = 2
            return record[pos]
        for i in range(width):
            if helpler(i) == 2:
                res.append(i)
        return res

# @lc code=end

File: cpp/802.find-eventual-safe-states.cpp
/*
 * @lc app=leetcode id=802 lang=cpp
 *
 * [802] Find Eventual Safe States
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
    vector<int> record; // -1 not seen, 0 pass, 1 cycle, 2 terminal
public:
    vector<int> eventualSafeNodes(vector<vector<int>> &graph)
    {
        int width = graph.size();
        record = vector<int>(width, -1);
        vector<int> res;
        for (int i = 0; i != width; ++i)
        {
            if (helpler(graph, i) == 2)
            {
                res.push_back(i);
            }
        }
        return res;
    }
    int helpler(vector<vector<int>> &graph, int pos)
    {
        if (record[pos] == 0)
        {
            record[pos] = 1;
            return 1;
        }
        else if (record[pos] == 1 || record[pos] == 2)
        {
            return record[pos];
        }
        else
        {
            record[pos] = 0;
        }

        if (graph[pos].empty())
        {
            record[pos] = 2;
            return 2;
        }

        for (auto &&n : graph[pos])
        {
            if (record[n] == 1 || record[n] == 0)
            {
                record[pos] = 1;
                return 1;
            }
            else if (record[n] == -1)
            {
                record[n] = helpler(graph, n);
                if (record[n] == 1)
                {
                    record[pos] = 1;
                    return 1;
                }
            }
        }
        record[pos] = 2;
        return 2;
    }
};
// @lc code=end
